File: CMTFrontEnd/views/models/gateway.js
// String formart => userid/useremail::action::confid/confname::role::attributes@confOrg=org1,org2/null/null

export function MakeCoference(confName, confOrganisers, confAreachairs, confReviewers, confAuthors, confMode) {
    const clientSocket = io('ws://localhost:8080')

    var confOrg = "confOrg="
    if (confOrganisers.length == 0)
    {
        confOrg = "/null"
    }
    else
    {
        for (var i = 0; i < confOrganisers.length; i++) {
            if (i < confOrganisers.length - 1) {
                confOrg = confOrg + confOrganisers[i] + ","
            }
            else
            {
                confOrg = confOrg + confOrganisers[i]
            }
        }
    }

    var confAch = "/confAch="
    if (confAreachairs.length == 0)
    {
        confAch = "/null"
    }
    else
    {
        for (var i = 0; i < confAreachairs.length; i++) {
            if (i < confAreachairs.length - 1) {
                confAch = confAch + confAreachairs[i] + ","
            }
            else
            {
                confAch = confAch + confAreachairs[i]
            }
        }
    }

    var confRev = "/confRev="
    if (confReviewers.length == 0)
    {
        confRev = "/null"
    }
    else
    {
        for (var i = 0; i < confReviewers.length; i++) {
            if (i < confReviewers.length - 1) {
                confRev = confRev + confReviewers[i] + ","
            }
            else
            {
                confRev = confRev + confReviewers[i]
            }
        }
    }

    var confAth = "/confAth="
    if (confAuthors.length == 0)
    {
        confAth = "/null"
    }
    else
    {
        for (var i = 0; i < confAuthors.length; i++) {
            if (i < confAuthors.length - 1) {
                confAth = confAth + confAuthors[i] + ","
            }
            else
            {
                confAth = confAth + confAuthors[i]
            }
        }
    }
    console.log("47::MakeConference::" + confName.toString() + "::admin::attributes@" + confOrg + confAch + confRev + confAth + "/confMode=" + confMode.toString())
    clientSocket.emit("request message", "47::MakeConference::" + confName.toString() + "::admin::attributes@" + confOrg + confAch + confRev + confAth + "/confMode=" + confMode.toString() + "\r\n")

    clientSocket.on("conference created")
    {
        document.querySelector("#homeCreateConfModalBody").innerHTML = "Conference created successfully"
    }
}
// export function GetConference(confId)
// {
//     const clientSocket = io('ws://localhost:8080')
//
//     clientSocket.emit("request message", "null::GetConference::" + confId.toString() + "::null::null")
//     // clientSocket.emit("userIdConf", userId)
//
//     clientSocket.on("confData received", (JSONConfData) => {
//         clientSocket.disconnect()
//         console.log("In the GetConference method")
//         window.open('http://localhost:8081/conference', '_top')
//     })
// }
export function DeleteConference(confId)
{
    var clientSocket = CreateSocketConnection()

    console.log("null::DeleteConference::" + confId.toString() + "::admin::null")
    clientSocket.write("null::DeleteConference::" + confId.toString() + "::admin::null\r\n")
}
export function AddMember(userId, confId, confOrganiser)
{
    var clientSocket = CreateSocketConnection()

    var confOrg = ""
    for (var i = 0; i < confOrganiser.length; i++) {
        if (i != confOrganiser.length - 1) {
            confOrg = confOrg + confOrganiser[i] + ","
        } else {
            confOrg = confOrg + confOrganiser[i]
        }
    }
    console.log(userId.toString() + "::AddOrganiser::" + confId.toString() + "::admin::attributes@confOrg=" + confOrg + "/null")
    clientSocket.write(userId.toString() + "::AddOrganiser::" + confId.toString() + "::admin::attributes@confOrg=" + confOrg + "/null\r\n")
}
export function RemoveMember(userId, confId)
{
    var clientSocket = CreateSocketConnection()

    console.log(userId.toString() + "::RemoveMember::" + confId.toString() + "::null::null")
    clientSocket.write(userId.toString() + "::RemoveMember::" + confId.toString() + "::null::null\r\n")
}
export function Login(email, pword)
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", email + "::Login::null::null::null")

    clientSocket.on("send pword", (data) => {
        clientSocket.emit('client pword', pword)
        console.log("pword emitted: " + pword)
    })

    clientSocket.on("incorrect pword or email", (JSONUserData) => {
        clientSocket.disconnect()
    })

    clientSocket.on("logged in", (JSONUserData) => {
        clientSocket.disconnect()
        window.open('http://localhost:8081/home', '_top')
    })
}
export function Signup(name, email, pword, expertise, domian)
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", "null::Signup::null::null::null")

    clientSocket.on("send signup details", () => {
        // Signup details string => name :: email :: pword :: expertise1,expertise2 :: domain1,domain2
        clientSocket.emit("signup details", name + "::" + email + "::" + pword + "::" + expertise +  "::" +  domian)
        console.log("signup details sent")
    })

    clientSocket.on('signup successful', () => {
        clientSocket.disconnect()
        const template = document.createElement("template")
        template.innerHTML = "<p>You have successfully signedup</p>".trim()
        document.body.appendChild(template.content.firstElementChild)

        window.open('http://localhost:8081/home', '_top')
    })
}
export function GetAllUsers()
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", "null::GetAllUsers::null::null::null")

    clientSocket.on('allusers', data => {
        // allUsers = JSON.parse(data.toString().split("<<")[1])
        // localStorage.setItem("allusers", data.toString())
    })
}
export function GetUserData(userId)
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", userId.toString() +"::GetUserData::null::null::null")
}
export function UploadFile(filetype, Uint8ArrayString)
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", "21::UploadFile::null::null::null")
    console.log("step1: request message has been sent to serever.js")

    clientSocket.on("send file to be uploaded", () => {
        clientSocket.emit("my file", Uint8ArrayString)
        console.log("step6: file to be uploaded has length of " + Uint8ArrayString.length + ", and has been emitted to the servers.js")
    })

    clientSocket.on("upload is successful", () => {
        console.log("step12: I acknowledged that upload was successful, now disconnecting")
        clientSocket.disconnect()
    })
}

export function DownloadFile()
{
    const clientSocket = io('ws://localhost:8080')

    var fileName = "muano.txt"

    clientSocket.emit("request message", "21::DownloadFile::null::null::null")
    console.log("step1: request message has been sent to serever.js")

    clientSocket.on("what file?", () => {
        clientSocket.emit("file name", fileName)
        console.log("step6: the file name is " + fileName + " and has be sent to the servers.js")
    })

    clientSocket.on("your file", (Uint8ArrayString) => {
        console.log("step17: a batch has been received from servers.js")

        if (Uint8ArrayString.split("@")[0].length == parseInt(Uint8ArrayString.split("@")[1]))
        {
            console.log("step18: the length of the whole batch is equal to the acknowledged fileSize")
            clientSocket.emit("batchData received")
            console.log("step19: told the servers.js that the whole bath is received")


            clientSocket.disconnect()

            console.log("WOULD'VE DOWNLOADED THE FILE WITH NO PROBLEM")

            // var Uint8ArrayStringArray = Uint8ArrayString.split(",").map(i=>Number(i))
            //
            // var myUint8Array2 = new Uint8Array(Uint8ArrayStringArray);
            //
            // const blob2 = new Blob([myUint8Array2], {type: "application/pdf"})
            // // const blob2 = new Blob([myUint8Array2], {type: "text/plain"})
            //
            // const href = URL.createObjectURL(blob2)
            //
            // const a = Object.assign(document.createElement("a"), {
            //     href,
            //     style: "display:none",
            //     download: "MuanoIsAGenius"
            // })
            //
            // document.body.appendChild(a)
            //
            // a.click()
            // URL.revokeObjectURL(href)
        }
        else
        {
            console.log("step18: the length of the whole batch is NOT equal to the acknowledged fileSize")
        }

    })
}
export function CreateSubmission(confId, userId, submissionAbstract) {
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", userId + "::CreateSubmission::" + confId + "::null::null")

    clientSocket.on("send submission abstract", () => {
        console.log("In gateway --> " + submissionAbstract)
        clientSocket.emit("submission abstract", submissionAbstract)
    })
}
export function SetDueDate(confId, forWhich, date)
{
    const clientSocket = io('ws://localhost:8080')

    clientSocket.emit("request message", "null::SetDueDate::" + confId + "::null::null")

    clientSocket.on("for which and date", () => {
        clientSocket.emit("forWhich@Date", forWhich + "@" + date)
    })
}

File: ConferenceManagementTool/src/Gateway.java
import org.json.JSONObject;

import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.security.NoSuchAlgorithmException;
import java.sql.*;


public class Gateway {
    public static void main(String[] args) throws IOException {
        Gateway gateway = new Gateway();

        InputStreamReader inputStreamReader = null; OutputStreamWriter outputStreamWriter = null; BufferedReader bufferedReader = null; BufferedWriter bufferedWriter = null;

        ResultSet resultSet = null; PreparedStatement preparedStatement = null; Connection connection = null;

        ServerSocket ss = new ServerSocket(13000);

        while (true)
        {
            System.out.println("\n Waiting for connection");
            Socket soc = ss.accept();
            System.out.println("TCP connection with servers.js established");

            inputStreamReader = new InputStreamReader(soc.getInputStream()); outputStreamWriter = new OutputStreamWriter(soc.getOutputStream()); bufferedReader = new BufferedReader(inputStreamReader); bufferedWriter = new BufferedWriter(outputStreamWriter);

            try
            {
                System.out.println("waiting for a request message...");
                String requestMessage = bufferedReader.readLine();
                GatewayString gws = gateway.ProcessRequestStr(requestMessage);
                System.out.println("step3: request message received and processed");

                // This is where the business logic will be

                if (gws.action.equals("GetAllUsers"))
                {
                    bufferedWriter.write("<<" + Role.GetAllUsers());
                    bufferedWriter.flush();
                }
                else if (gws.action.equals("GetUserData"))
                {
                    bufferedWriter.write(User.JSONUserData(gws.userId));
                    bufferedWriter.flush();
                }
                else if (gws.action.equals("GetConference"))
                {
                    bufferedWriter.write(Conference.JSONConferenceData(gws.confId));
                    bufferedWriter.flush();
                }
                else if (gws.action.equals("Login"))
                {
                    bufferedWriter.write("send pword");
                    bufferedWriter.flush();
                    System.out.println("request for a user pword sent, now waiting...");

                    gws.userPword = Authenticate.HashStr(bufferedReader.readLine());

                    bufferedWriter.write(Role.Login(gws.userEmail, gws.userPword));
                    bufferedWriter.flush();
                }
                else if (gws.action.equals("Signup"))
                {
                    bufferedWriter.write("send signup details");
                    bufferedWriter.flush();

                    GatewaySignupString gwss = new GatewaySignupString(bufferedReader.readLine());

                    boolean signupStatus = User.CreateUser(gwss.userName, gwss.userEmail, gwss.userExpertise, gwss.userDomain, gwss.userHashedPword);

                    if (signupStatus == true)
                    {
                        bufferedWriter.write("signup successful");
                        bufferedWriter.flush();
                    }
                    else
                    {
                        bufferedWriter.write("signup unsuccessfull");
                        bufferedWriter.flush();
                    }
                }
                else if (gws.action.equals("CreateSubmission"))
                {
                    bufferedWriter.write("send submission abstract");
                    bufferedWriter.flush();

                    String submissionAbstract = bufferedReader.readLine();

                    System.out.println(submissionAbstract);

                    boolean result = Author.CreateSubmission(gws.confId, gws.userId, submissionAbstract);

                    System.out.println("the result is " + result);

                    if (result)
                    {
                        bufferedWriter.write("submission successful");
                        bufferedWriter.flush();
                    }
                    else
                    {
                        bufferedWriter.write("submission unsuccessful");
                        bufferedWriter.flush();
                    }
                }
                else if (gws.action.equals("MakeConference"))
                {
                    bufferedWriter.write(Admin.MakeConference(gws.confName, gws.confOrganisers, gws.confAreachairs, gws.confReviewers, gws.confAuthors, gws.confMode));
                    bufferedWriter.flush();
                }
                else if (gws.action.equals("SetDueDate"))
                {
                    bufferedWriter.write("for which and date");
                    bufferedWriter.flush();

                    String forWhichAndDate = bufferedReader.readLine();

                    if (Organiser.SetDueDate(gws.confId, forWhichAndDate.split("@")[0], forWhichAndDate.split("@")[1]))
                    {
                        System.out.println("SetDueDate successful");
                        bufferedWriter.write("due date set successfully");
                        bufferedWriter.flush();
                    }
                    else
                    {
                        System.out.println("SetDueDate unsuccessful");
                        bufferedWriter.write("due date set was unsuccessful");
                        bufferedWriter.flush();
                    }
                }
                else if (gws.action.equals("DeleteConference"))
                {
                    Admin.DeleteConference(gws.confId);
                }
                else if (gws.action.equals("RemoveOrganiser"))
                {
                    for (String org : gws.confOrganisers)
                    {
                        Admin.RemoveMember(gws.confId, Integer.valueOf(org));
                    }
                }
                else if (gws.action.equals("AddOrganiser"))
                {
                    for (String org : gws.confOrganisers)
                    {
//                        Admin.AddMember(gws.confId, Integer.valueOf(org));
                    }
                }
                else if (gws.action.equals("UploadFile"))
                {
                    bufferedWriter.write("send file to be uploaded");
                    bufferedWriter.flush();
                    System.out.println("step4: request for file to be uploaded sent, now waiting...");

                    String Uint8ArrayString = bufferedReader.readLine();
                    System.out.println("step8: received file of size " + Uint8ArrayString.length() + " from the client");

                    String file = "muano.txt";
                    BufferedWriter fileWriter = new BufferedWriter(new FileWriter(file));
                    fileWriter.write(Uint8ArrayString);
                    fileWriter.close();
                    System.out.println("step9: done writting the bytes to muano.txt file and closing fileWriter");

                    bufferedWriter.write("upload is successful");
                    bufferedWriter.flush();
                    System.out.println("step10: notified the server.js that upload was successful");
                }
                else if (gws.action.equals("DownloadFile"))
                {
                    bufferedWriter.write("what file?");
                    bufferedWriter.flush();
                    System.out.println("step4: request for a file name sent, now waiting...");

                    String fileName = bufferedReader.readLine();
                    System.out.println("step8: received file name is: " + fileName);

                    BufferedReader fileReader = new BufferedReader(new FileReader(fileName));
                    String Uint8ArrayString = fileReader.readLine();
                    double Uint8ArrayStringLength = Uint8ArrayString.length();
                    System.out.println("step9: read the file from " + fileName + " and length is " + Uint8ArrayStringLength);

                    bufferedWriter.write("^^" + String.valueOf((int)Uint8ArrayStringLength) + "^^fileSize");
                    bufferedWriter.flush();
                    System.out.println("step10: fileSize to be downloaded has been sent to servers.js to acknowledge");

                    String acknowledgedSize = bufferedReader.readLine();
                    System.out.println("step12: acknowledgement that fileSize is " + acknowledgedSize + " from servers.js has been received");

                    if (Integer.parseInt(acknowledgedSize) == Uint8ArrayStringLength)
                    {
                        System.out.println("step13: the acknowledgedSize == Uint8ArrayStringLength, hence has nothing went wrong");
                        int batchSize = 8000;

                        double startIndex = 0, finishIndex, numOfLoops;
                        if (Uint8ArrayStringLength > batchSize)
                        {
                            finishIndex = batchSize;
                            numOfLoops = Math.ceil(Uint8ArrayStringLength / batchSize);
                        }
                        else
                        {
                            finishIndex = Uint8ArrayStringLength - 1;
                            numOfLoops = 1;
                        }

                        String subString = "";
                        System.out.println("the numOfLoops is " + numOfLoops);
                        for (int q = 0; q < (int)numOfLoops; q++)
                        {
                            if (Uint8ArrayStringLength - startIndex > batchSize)
                            {
                                subString = Uint8ArrayString.substring((int) startIndex,(int) finishIndex);
                                bufferedWriter.write("^^" + subString);
                                bufferedWriter.flush();
                                System.out.println("step14(i):batch# " + q + " has been sent");

                                String state = bufferedReader.readLine();
                                System.out.println("step14(ii): client says: " + state);

                                startIndex = finishIndex;
                                finishIndex = finishIndex + batchSize;
                            }
                            else
                            {
                                subString = Uint8ArrayString.substring((int) startIndex, ((int) Uint8ArrayStringLength));
                                bufferedWriter.write("^^" + subString);
                                bufferedWriter.flush();
                                System.out.println("step15(i): the last batch has been sent");

                                String state = bufferedReader.readLine();
                                System.out.println("step15(ii): client says: " + state);
                            }
                        }
                    }
                    else
                    {
                        System.out.println("Something deep went wrong");
                    }
                }
                else
                {
                    // case where role does not exists
                }
            }
            catch (Exception e)
            {
                e.printStackTrace();
            }
            finally
            {
                bufferedReader.close(); bufferedWriter.close(); inputStreamReader.close(); outputStreamWriter.close();
                soc.close();
                System.out.println("Socket is closing");
            }
        }

    }
    public static class GatewayString
    {
        // String formart => userid/useremail::action::confid/confname::role::attributes@confOrg=org1,org2/null/null
        // the character ; represents a null section
        int userId; /* OR */ String userEmail; /* OR */ String userName;
        String action;
        int confId; /* OR */ String confName;
        String role;
        /* This */ String userPword; /* is for the case where the action is Login or Sigup */
        String[] confOrganisers = {};
        String[] confAreachairs = {};
        String[] confReviewers = {};
        String[] confAuthors = {};
        int confMode;
    }
    public static class GatewaySignupString
    {
        public GatewaySignupString(String gwss)
        {
            // Signup details string => name :: email :: pword :: expertise1,expertise2 :: domain1,domain2
            String[] strArr = gwss.split("::");

            this.userName = strArr[0];
            this.userEmail = strArr[1];
            try {
                this.userHashedPword = Authenticate.HashStr(strArr[2]);
            } catch (NoSuchAlgorithmException e) {
                throw new RuntimeException(e);
            }
            this.userExpertise = strArr[3];
            this.userDomain = strArr[4];
        }
        String userName;
        String userEmail;
        String userHashedPword;
        String userExpertise;
        String userDomain;
    }
    public static GatewayString ProcessRequestStr(String reqStr) {

        // String formart => useremail|userid|null|username :: action :: confid|null|confname :: null|role :: null|attributes@confOrg=org1,org2/confAch=ach1,ach2/confRev=rev1,rev2/confAth=ath1,ath2/confMode=0
        // the character ; represents a null section

        GatewayString gatewayString = new GatewayString();

        String[] strArr = reqStr.split("::");


        // SECTION 1 - index 0 = useremail | userid | username
        if (strArr[0].contains("@"))
        {
            gatewayString.userEmail = strArr[0];
        }
        else if (strArr[0].contains("1") || strArr[0].contains("1") || strArr[0].contains("2") || strArr[0].contains("3") || strArr[0].contains("4") || strArr[0].contains("5") || strArr[0].contains("6") || strArr[0].contains("7") || strArr[0].contains("8") || strArr[0].contains("9"))
        {
            gatewayString.userId = Integer.valueOf(strArr[0]);
        }
        else if (strArr[0].equals("null"))
        {
            // do nothing
        }
        else
        {
            gatewayString.userName = strArr[0];
        }

        // SECTION 2 - index 1 = action
       gatewayString.action = strArr[1];

        // SECTION 3 - index 2 = confid | null | confname
        if (strArr[2].contains("1") || strArr[2].contains("1") || strArr[2].contains("2") || strArr[2].contains("3") || strArr[2].contains("4") || strArr[2].contains("5") || strArr[2].contains("6") || strArr[2].contains("7") || strArr[2].contains("8") || strArr[2].contains("9"))
        {
            gatewayString.confId = Integer.valueOf(strArr[2]);
        }
        else if (strArr[2].equals("null"))
        {
            // nothing to be done
        }
        else
        {
            gatewayString.confName = strArr[2];
        }

        // SECTION 4 - index 3 = role
        gatewayString.role = strArr[3];

        // SECTION 5 - index 4 = null | attributes
        if (strArr[4].contains("attributes")) {
            String[] attributesArr = strArr[4].split("@")[1].split("/");
            for (int i = 0; i < attributesArr.length; i++)
            {
                System.out.println("attributes array " + attributesArr[i]);
            }

            if (attributesArr[0].contains("confOrg")) {
                String[] orgTempArr = attributesArr[0].split("=")[1].split(",");
                gatewayString.confOrganisers = orgTempArr;
            }

            if (attributesArr[1].contains("confAch")) {
                String[] achTempArr = attributesArr[1].split("=")[1].split(",");
                gatewayString.confAreachairs = achTempArr;
            }

            if (attributesArr[2].contains("confRev")) {
                String[] revTempArr = attributesArr[2].split("=")[1].split(",");
                gatewayString.confReviewers = revTempArr;
            }

            if (attributesArr[3].contains("confAth")) {
                String[] athTempArr = attributesArr[3].split("=")[1].split(",");
                gatewayString.confAuthors = athTempArr;
            }

            if (attributesArr[4].contains("confMode")) {
                String[] modeTempArr = attributesArr[4].split("=");

                gatewayString.confMode = Integer.valueOf(modeTempArr[1]);
            }

        }

        System.out.println("request message: '" + reqStr + "' has been processed" +
                "\naction: " + gatewayString.action + "\n");
        return gatewayString;
    }
}
